### PR TITLE
persist: change downcast_stats error! log to warn!

### PR DESCRIPTION
**Pull request by danhhz**
We saw this pop up, which means the error! has done its job. While we investigate, downgrade the log to a warn! to avoid spamming sentry. Add metrics coverage to compensate, so we don't have to rely on searching logs to see if it's still happening.

While I'm in here, add some context to the message so it's easier to track down next time (I had to do a lot of manual correlating during the initial triage).

Touches #22680

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by guswynn — APPROVED** [review]
nice

File: src/persist-client/src/internal/metrics.rs [hunk 1]
```diff
@@ -2026,6 +2027,10 @@ impl PushdownMetrics {
                 name: "mz_persist_pushdown_parts_stats_trimmed_bytes",
                 help: "total bytes trimmed from part stats",
             )),
+            parts_mismatched_stats_count: registry.register(metric!(
```
Comment (bkirwi) [inline src/persist-client/src/internal/metrics.rs:2030]: 🙇

**Review by bkirwi — APPROVED** [review]
Glad to have the extra debugging info! I'd probably default to leaving this at `error!`, but happy to downgrade if you prefer it.

File: src/storage-operators/src/persist_source.rs [hunk 2]
```diff
@@ -466,19 +473,33 @@ impl PendingWork {
...
+    stats: &'a dyn DynStats,
+) -> Option<&'a T::Stats> {
     match stats.as_any().downcast_ref::<T::Stats>() {
         Some(x) => Some(x),
         None => {
-            error!(
-                "unexpected stats type for {}: {}",
+            // TODO: There is a known instance of this #22680. While we look
+            // into it, log at warn instead of error to avoid spamming Sentry.
+            // Once we fix it, flip this back to error.
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/storage-operators/src/persist_source.rs:493]: I do prefer it, yeah, for three reasons:
- All 38 of those happened during rehydration the last time the affected env restarted. We'll keep getting sentry pings every time it restarts until whatever this is is fixed, which might concern someone (QA, maybe) and cause them to spend time looking into it before finding out that it's known and ~fine
- Our logging guide intends that `error!` should be something that pages the oncall, which this definitely doesn't meet the bar for at the moment
- The potential volume of these being sent to sentry if e.g. the affected env happens to get stuck in an OOM loop

**Review by danhhz — COMMENTED** [review]
TFTRs!

**Pull request merged**